File: trading/copytrade.py
import os
import sys
import logging
import asyncio
from typing import Dict, List, Any, Optional
import time

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from config import WETH_ADDRESS
from database import (
    get_user_by_telegram_id, create_copy_trade_config, 
    update_copy_trade_config, get_user_copy_trade_configs
)
from utils.wallet import decrypt_private_key
from trading.swap import buy_token_with_eth, sell_token_for_eth
from web3_module.web3_client import web3_client

# Setup logging
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# In-memory store for monitored wallets and their transactions
# In a production environment, this would be backed by a database
monitored_wallets = {}
processed_transactions = set()

async def setup_copy_trading(telegram_id: str, target_wallet: str, slippage_guard: float = 5.0, 
                            max_eth_per_trade: float = 0.1, sandbox_mode: bool = True) -> Dict[str, Any]:
    """
    Set up copy trading for a user
    
    Args:
        telegram_id: User's Telegram ID
        target_wallet: Address of the wallet to copy trades from
        slippage_guard: Maximum allowed slippage percentage
        max_eth_per_trade: Maximum ETH amount per trade
        sandbox_mode: Whether to run in sandbox mode (no real trades)
        
    Returns:
        Dictionary containing setup result
    """
    try:
        # Validate target wallet address
        if not web3_client.is_address(target_wallet):
            return {
                "error": "Invalid wallet address",
                "message": "The provided wallet address is not valid."
            }
        
        # Get user from database
        user = get_user_by_telegram_id(telegram_id)
        if not user or not user.wallet_address:
            return {
                "error": "No wallet found",
                "message": "You don't have a wallet set up yet. Use /wallet to set one up."
            }
        
        # Create copy trade config
        config = create_copy_trade_config(
            telegram_id=telegram_id,
            target_wallet=target_wallet,
            slippage_guard=slippage_guard,
            active=True,
            sandbox_mode=sandbox_mode,
            max_eth_per_trade=max_eth_per_trade
        )
        
        # Add to monitored wallets
        if target_wallet not in monitored_wallets:
            monitored_wallets[target_wallet] = []
        
        monitored_wallets[target_wallet].append({
            "telegram_id": telegram_id,
            "config_id": config.id,
            "slippage_guard": slippage_guard,
            "max_eth_per_trade": max_eth_per_trade,
            "sandbox_mode": sandbox_mode
        })
        
        return {
            "success": True,
            "config_id": config.id,
            "target_wallet": target_wallet,
            "slippage_guard": slippage_guard,
            "max_eth_per_trade": max_eth_per_trade,
            "sandbox_mode": sandbox_mode,
            "message": "Copy trading set up successfully!"
        }
    
    except Exception as e:
        logger.error(f"Error setting up copy trading: {str(e)}")
        return {
            "error": "Error setting up copy trading",
            "message": f"An error occurred: {str(e)}"
        }

async def update_copy_trading(telegram_id: str, config_id: int, active: Optional[bool] = None,
                             slippage_guard: Optional[float] = None, sandbox_mode: Optional[bool] = None,
                             max_eth_per_trade: Optional[float] = None) -> Dict[str, Any]:
    """
    Update copy trading configuration
    
    Args:
        telegram_id: User's Telegram ID
        config_id: ID of the copy trade configuration
        active: Whether the configuration is active
        slippage_guard: Maximum allowed slippage percentage
        sandbox_mode: Whether to run in sandbox mode (no real trades)
        max_eth_per_trade: Maximum ETH amount per trade
        
    Returns:
        Dictionary containing update result
    """
    try:
        # Get user from database
        user = get_user_by_telegram_id(telegram_id)
        if not user:
            return {
                "error": "User not found",
                "message": "User not found in database."
            }
        
        # Update copy trade config
        config = update_copy_trade_config(
            config_id=config_id,
            active=active,
            slippage_guard=slippage_guard,
            sandbox_mode=sandbox_mode,
            max_eth_per_trade=max_eth_per_trade
        )
        
        if not config:
            return {
                "error": "Configuration not found",
                "message": "The specified copy trade configuration was not found."
            }
        
        # Update monitored wallets
        for wallet, configs in monitored_wallets.items():
            for i, cfg in enumerate(configs):
                if cfg["config_id"] == config_id:
                    if active is not None:
                        if not active:
                            # Remove from monitored wallets if deactivated
                            monitored_wallets[wallet].pop(i)
                            if not monitored_wallets[wallet]:
                                del monitored_wallets[wallet]
                        else:
                            cfg["active"] = active
                    
                    if slippage_guard is not None:
                        cfg["slippage_guard"] = slippage_guard
                    
                    if sandbox_mode is not None:
                        cfg["sandbox_mode"] = sandbox_mode
                    
                    if max_eth_per_trade is not None:
                        cfg["max_eth_per_trade"] = max_eth_per_trade
                    
                    break
        
        return {
            "success": True,
            "config_id": config_id,
            "active": active if active is not None else config.active,
            "slippage_guard": slippage_guard if slippage_guard is not None else config.slippage_guard,
            "sandbox_mode": sandbox_mode if sandbox_mode is not None else config.sandbox_mode,
            "max_eth_per_trade": max_eth_per_trade if max_eth_per_trade is not None else config.max_eth_per_trade,
            "message": "Copy trading configuration updated successfully!"
        }
    
    except Exception as e:
        logger.error(f"Error updating copy trading: {str(e)}")
        return {
            "error": "Error updating copy trading",
            "message": f"An error occurred: {str(e)}"
        }

async def get_copy_trading_configs(telegram_id: str) -> Dict[str, Any]:
    """
    Get copy trading configurations for a user
    
    Args:
        telegram_id: User's Telegram ID
        
    Returns:
        Dictionary containing copy trading configurations
    """
    try:
        # Get user from database
        user = get_user_by_telegram_id(telegram_id)
        if not user:
            return {
                "error": "User not found",
                "message": "User not found in database."
            }
        
        # Get copy trade configs
        configs = get_user_copy_trade_configs(telegram_id)
        
        # Format configs
        formatted_configs = []
        for config in configs:
            formatted_config = {
                "id": config.id,
                "target_wallet": config.target_wallet,
                "slippage_guard": config.slippage_guard,
                "active": config.active,
                "sandbox_mode": config.sandbox_mode,
                "max_eth_per_trade": config.max_eth_per_trade,
                "created_at": config.created_at.strftime("%Y-%m-%d %H:%M:%S"),
                "updated_at": config.updated_at.strftime("%Y-%m-%d %H:%M:%S")
            }
            formatted_configs.append(formatted_config)
        
        return {
            "configs": formatted_configs,
            "count": len(formatted_configs)
        }
    
    except Exception as e:
        logger.error(f"Error getting copy trading configs: {str(e)}")
        return {
            "error": "Error getting copy trading configs",
            "message": f"An error occurred: {str(e)}"
        }

async def process_transaction(transaction: Dict[str, Any]) -> None:
    """
    Process a transaction for copy trading
    
    Args:
        transaction: Transaction data
    """
    try:
        # Check if transaction is already processed
        tx_hash = transaction["hash"]
        if tx_hash in processed_transactions:
            return
        
        # Add to processed transactions
        processed_transactions.add(tx_hash)
        
        # Get transaction details
        from_address = transaction["from"]
        to_address = transaction["to"]
        value = transaction["value"]
        
        # Check if this is a transaction from a monitored wallet
        if from_address in monitored_wallets:
            # This is a transaction from a monitored wallet
            logger.info(f"Detected transaction from monitored wallet: {from_address}")
            
            # Get the users who are copying this wallet
            copying_users = monitored_wallets[from_address]
            
            # Process the transaction for each user
            for user_config in copying_users:
                telegram_id = user_config["telegram_id"]
                slippage_guard = user_config["slippage_guard"]
                max_eth_per_trade = user_config["max_eth_per_trade"]
                sandbox_mode = user_config["sandbox_mode"]
                
                # Get user from database
                user = get_user_by_telegram_id(telegram_id)
                if not user or not user.wallet_address or not user.encrypted_private_key:
                    logger.warning(f"User {telegram_id} has no wallet or private key")
                    continue
                
                # In a real implementation, you would:
                # 1. Detect if this is a buy or sell transaction
                # 2. Identify the token being bought or sold
                # 3. Calculate the amount to buy or sell based on the user's settings
                # 4. Execute the trade
                
                # For now, we'll just log the transaction
                logger.info(f"Would copy trade for user {telegram_id}: {tx_hash}")
                
                if not sandbox_mode:
                    logger.info(f"Real trading not implemented yet")
                
        # In a real implementation, you would also monitor DEX router contracts
        # to detect swaps and copy them
    
    except Exception as e:
        logger.error(f"Error processing transaction: {str(e)}")

async def start_monitoring() -> None:
    """
    Start monitoring for transactions to copy
    """
    logger.info("Starting copy trade monitoring")
    
    # In a real implementation, you would:
    # 1. Subscribe to new blocks using web3.eth.filter('latest')
    # 2. For each new block, get all transactions
    # 3. Filter transactions from monitored wallets
    # 4. Process those transactions for copy trading
    
    # For now, we'll just simulate monitoring with a simple loop
    while True:
        try:
            # Get latest block number
            latest_block = web3_client.w3.eth.block_number
            logger.debug(f"Latest block: {latest_block}")
            
            # In a real implementation, you would process transactions here
            
            # Sleep for a bit
            await asyncio.sleep(5)
        
        except Exception as e:
            logger.error(f"Error in monitoring loop: {str(e)}")
            await asyncio.sleep(10)  # Sleep a bit longer on error 

File: services/copytrade.js
/**
 * Copy Trade Service for Zoracle Telegram Bot
 * 
 * This service monitors transactions from target wallets and mirrors their trades
 * based on user configurations.
 */
const { ethers } = require('ethers');
const { EventEmitter } = require('events');
const { Wallet, Contract } = require('ethers');
const { formatEther, parseEther, formatUnits } = require('ethers/lib/utils');
const { CONFIG, ABIS } = require('../config');
const { CopyTradeOps, UserOps, TransactionOps } = require('../database/operations');
const { buyTokensWithEth, sellTokensForEth, getTokenPrice } = require('./trading');
const mempoolMonitor = require('./mempoolMonitor');

// Create event emitter for trade events
const copyTradeEvents = new EventEmitter();

// Track active copy trade monitors
const activeMonitors = new Map();

/**
 * Initialize copy trade service
 * @returns {Promise<boolean>} - True if initialized successfully
 */
async function initialize() {
  try {
    console.log('🔄 Initializing copy trade service...');
    
    // Start mempool monitoring if not already active
    if (!mempoolMonitor.isActive()) {
      mempoolMonitor.startMonitoring();
    }
    
    // Get all active copy trades from database
    const activeCopyTrades = await getAllActiveCopyTrades();
    
    // Start monitoring for each target wallet
    for (const copyTrade of activeCopyTrades) {
      await startMonitoring(copyTrade.targetWallet);
    }
    
    console.log(`✅ Copy trade service initialized with ${activeCopyTrades.length} active monitors`);
    return true;
  } catch (error) {
    console.error('❌ Failed to initialize copy trade service:', error.message);
    return false;
  }
}

/**
 * Get all active copy trades from database
 * @returns {Promise<Array>} - Array of active copy trades
 */
async function getAllActiveCopyTrades() {
  try {
    // Query database for all active copy trades
    // Group by target wallet to avoid duplicate monitors
    const copyTrades = await CopyTradeOps.getAllActiveCopyTrades();
    
    // Group by target wallet
    const groupedByTarget = copyTrades.reduce((acc, copyTrade) => {
      const targetWallet = copyTrade.targetWallet.toLowerCase();
      if (!acc[targetWallet]) {
        acc[targetWallet] = [];
      }
      acc[targetWallet].push(copyTrade);
      return acc;
    }, {});
    
    // Flatten to unique target wallets with their followers
    return Object.keys(groupedByTarget).map(targetWallet => ({
      targetWallet,
      followers: groupedByTarget[targetWallet]
    }));
  } catch (error) {
    console.error('❌ Failed to get active copy trades:', error.message);
    return [];
  }
}

/**
 * Start monitoring a target wallet
 * @param {string} targetWallet - Target wallet address
 * @returns {Promise<boolean>} - True if started successfully
 */
async function startMonitoring(targetWallet) {
  try {
    // Normalize wallet address
    const normalizedAddress = targetWallet.toLowerCase();
    
    // Check if already monitoring
    if (activeMonitors.has(normalizedAddress)) {
      console.log(`🔍 Already monitoring wallet: ${normalizedAddress}`);
      return true;
    }
    
    console.log(`🔍 Starting copy trade monitoring for wallet: ${normalizedAddress}`);
    
    // Add mempool listener for this wallet
    const listenerId = `copy_trade_${normalizedAddress}`;
    
    mempoolMonitor.addListener(
      listenerId,
      'copy_trade',
      (tx) => tx.from.toLowerCase() === normalizedAddress,
      async (tx, data) => {
        await processCopyTrade(tx, data);
      },
      { targetWallet: normalizedAddress }
    );
    
    // Store in active monitors
    activeMonitors.set(normalizedAddress, {
      listenerId,
      startTime: Date.now()
    });
    
    console.log(`✅ Copy trade monitoring started for wallet: ${normalizedAddress}`);
    return true;
  } catch (error) {
    console.error(`❌ Failed to start monitoring for ${targetWallet}:`, error.message);
    return false;
  }
}

/**
 * Stop monitoring a target wallet
 * @param {string} targetWallet - Target wallet address
 * @returns {Promise<boolean>} - True if stopped successfully
 */
async function stopMonitoring(targetWallet) {
  try {
    // Normalize wallet address
    const normalizedAddress = targetWallet.toLowerCase();
    
    // Check if monitoring
    if (!activeMonitors.has(normalizedAddress)) {
      console.log(`🔍 Not monitoring wallet: ${normalizedAddress}`);
      return true;
    }
    
    console.log(`🔍 Stopping copy trade monitoring for wallet: ${normalizedAddress}`);
    
    // Get listener ID
    const { listenerId } = activeMonitors.get(normalizedAddress);
    
    // Remove mempool listener
    mempoolMonitor.removeListener(listenerId);
    
    // Remove from active monitors
    activeMonitors.delete(normalizedAddress);
    
    console.log(`✅ Copy trade monitoring stopped for wallet: ${normalizedAddress}`);
    return true;
  } catch (error) {
    console.error(`❌ Failed to stop monitoring for ${targetWallet}:`, error.message);
    return false;
  }
}

/**
 * Process a copy trade transaction
 * @param {Object} tx - Transaction object
 * @param {Object} data - Additional data
 * @returns {Promise<void>}
 */
async function processCopyTrade(tx, data) {
  try {
    const { targetWallet, copyTrades } = data;
    
    // Skip if no copy trades
    if (!copyTrades || copyTrades.length === 0) return;
    
    console.log(`👥 Processing copy trade transaction: ${tx.hash}`);
    console.log(`   Target wallet: ${targetWallet}`);
    console.log(`   Number of followers: ${copyTrades.length}`);
    
    // Analyze transaction to determine if it's a trade
    const tradeInfo = await analyzeTrade(tx);
    
    // Skip if not a trade
    if (!tradeInfo) {
      console.log(`   Not a trade transaction, skipping`);
      return;
    }
    
    console.log(`   Trade type: ${tradeInfo.type}`);
    console.log(`   Token: ${tradeInfo.tokenAddress}`);
    console.log(`   Amount: ${tradeInfo.amount}`);
    
    // Emit trade event
    copyTradeEvents.emit('trade', {
      tx,
      targetWallet,
      tradeInfo,
      copyTrades
    });
    
    // Execute mirror trades for each follower
    for (const copyTrade of copyTrades) {
      await executeMirrorTrade(copyTrade, tradeInfo);
    }
  } catch (error) {
    console.error(`❌ Error processing copy trade for ${tx.hash}:`, error.message);
  }
}

/**
 * Analyze a transaction to determine if it's a trade
 * @param {Object} tx - Transaction object
 * @returns {Promise<Object|null>} - Trade info or null if not a trade
 */
async function analyzeTrade(tx) {
  try {
    // Skip transactions with no data
    if (!tx.data || tx.data === '0x') return null;
    
    // Check if transaction is to a DEX router
    const isDexInteraction = Object.values(CONFIG.DEX_ROUTERS).some(
      router => tx.to && tx.to.toLowerCase() === router.toLowerCase()
    );
    
    if (!isDexInteraction) return null;
    
    // Try to decode transaction data
    const aerodromeRouterInterface = new ethers.utils.Interface(ABIS.AERODROME_ROUTER_ABI);
    
    try {
      const decodedData = aerodromeRouterInterface.parseTransaction({ data: tx.data, value: tx.value });
      const { name: functionName, args } = decodedData;
      
      // Process based on function name
      if (functionName === 'swapExactETHForTokens') {
        // ETH to Token swap (Buy)
        return {
          type: 'buy',
          tokenAddress: args.route[0].output,
          amount: formatEther(tx.value),
          ethAmount: formatEther(tx.value),
          rawAmount: tx.value,
          decodedData
        };
      } else if (functionName === 'swapExactTokensForETH') {
        // Token to ETH swap (Sell)
        return {
          type: 'sell',
          tokenAddress: args.route[0].input,
          amount: formatUnits(args.amountIn, 18), // Assuming 18 decimals, should be fetched
          ethAmount: null, // Will be determined after execution
          rawAmount: args.amountIn,
          decodedData
        };
      } else if (functionName === 'swapExactTokensForTokens') {
        // Token to Token swap (Sell + Buy)
        return {
          type: 'token_to_token',
          inputToken: args.route[0].input,
          outputToken: args.route[args.route.length - 1].output,
          amount: formatUnits(args.amountIn, 18), // Assuming 18 decimals
          rawAmount: args.amountIn,
          decodedData
        };
      }
    } catch (error) {
      console.error(`Error decoding transaction data: ${error.message}`);
      return null;
    }
    
    return null;
  } catch (error) {
    console.error(`Error analyzing trade: ${error.message}`);
    return null;
  }
}

/**
 * Execute a mirror trade for a follower
 * @param {Object} copyTrade - Copy trade configuration
 * @param {Object} tradeInfo - Trade information
 * @returns {Promise<Object|null>} - Transaction result or null if failed
 */
async function executeMirrorTrade(copyTrade, tradeInfo) {
  try {
    console.log(`👥 Executing mirror trade for user ${copyTrade.telegramId}`);
    
    // Get user data
    const user = await UserOps.getUser(copyTrade.telegramId);
    
    if (!user || !user.walletAddress || !user.encryptedPrivateKey) {
      console.error(`❌ User ${copyTrade.telegramId} has no wallet`);
      return null;
    }
    
    // Add random delay for MEV protection
    const delay = Math.floor(
      Math.random() * (CONFIG.COPY_TRADING.MAX_DELAY - CONFIG.COPY_TRADING.MIN_DELAY) + 
      CONFIG.COPY_TRADING.MIN_DELAY
    );
    
    console.log(`   Adding ${delay}ms delay for MEV protection`);
    await new Promise(resolve => setTimeout(resolve, delay));
    
    // Execute trade based on type
    if (tradeInfo.type === 'buy') {
      return await executeMirrorBuy(user, copyTrade, tradeInfo);
    } else if (tradeInfo.type === 'sell') {
      return await executeMirrorSell(user, copyTrade, tradeInfo);
    } else if (tradeInfo.type === 'token_to_token') {
      // For token to token swaps, we'll skip for now
      console.log(`   Token to token swaps not supported yet`);
      return null;
    }
    
    return null;
  } catch (error) {
    console.error(`❌ Error executing mirror trade:`, error.message);
    return null;
  }
}

/**
 * Execute a mirror buy
 * @param {Object} user - User data
 * @param {Object} copyTrade - Copy trade configuration
 * @param {Object} tradeInfo - Trade information
 * @returns {Promise<Object|null>} - Transaction result or null if failed
 */
async function executeMirrorBuy(user, copyTrade, tradeInfo) {
  try {
    // Calculate ETH amount based on copy trade settings
    const targetEthAmount = parseFloat(tradeInfo.ethAmount);
    const maxEthPerTrade = parseFloat(copyTrade.maxEthPerTrade);
    
    // Use the smaller of the two amounts
    let ethAmount = Math.min(targetEthAmount, maxEthPerTrade);
    
    // Check user's ETH balance
    const wallet = new ethers.Wallet(user.encryptedPrivateKey, new ethers.providers.JsonRpcProvider(CONFIG.PROVIDER_URL));
    const ethBalance = parseFloat(formatEther(await wallet.getBalance()));
    
    // Ensure we have enough ETH (leave some for gas)
    ethAmount = Math.min(ethAmount, ethBalance * 0.95);
    
    if (ethAmount <= 0) {
      console.error(`❌ Insufficient ETH balance for mirror trade`);
      return null;
    }
    
    console.log(`   Mirroring buy: ${ethAmount} ETH for ${tradeInfo.tokenAddress}`);
    
    // If in sandbox mode, just simulate the trade
    if (copyTrade.sandboxMode) {
      console.log(`   Sandbox mode: Simulating trade`);
      
      // Record simulated transaction
      await TransactionOps.createTransaction({
        telegramId: user.telegramId,
        type: 'simulated_buy',
        tokenAddress: tradeInfo.tokenAddress,
        amount: '0', // Simulated
        ethValue: ethAmount.toString(),
        status: 'simulated',
        timestamp: new Date()
      });
      
      // Emit event for notification
      copyTradeEvents.emit('mirror_trade_simulated', {
        telegramId: user.telegramId,
        type: 'buy',
        tokenAddress: tradeInfo.tokenAddress,
        ethAmount: ethAmount.toString()
      });
      
      return {
        simulated: true,
        ethAmount: ethAmount.toString()
      };
    }
    
    // Execute real trade
    const privateKey = user.encryptedPrivateKey; // In a real implementation, this should be decrypted
    
    const result = await buyTokensWithEth(
      tradeInfo.tokenAddress,
      ethAmount.toString(),
      privateKey,
      copyTrade.slippage,
      user.telegramId
    );
    
    // Emit event for notification
    copyTradeEvents.emit('mirror_trade_executed', {
      telegramId: user.telegramId,
      type: 'buy',
      tokenAddress: tradeInfo.tokenAddress,
      ethAmount: ethAmount.toString(),
      tokenAmount: result.tokenAmount,
      txHash: result.txHash
    });
    
    return result;
  } catch (error) {
    console.error(`❌ Error executing mirror buy:`, error.message);
    return null;
  }
}

/**
 * Execute a mirror sell
 * @param {Object} user - User data
 * @param {Object} copyTrade - Copy trade configuration
 * @param {Object} tradeInfo - Trade information
 * @returns {Promise<Object|null>} - Transaction result or null if failed
 */
async function executeMirrorSell(user, copyTrade, tradeInfo) {
  try {
    // Check if user has the token
    const provider = new ethers.providers.JsonRpcProvider(CONFIG.PROVIDER_URL);
    const tokenContract = new Contract(tradeInfo.tokenAddress, ABIS.ERC20_ABI, provider);
    
    // Get token decimals
    const decimals = await tokenContract.decimals();
    
    // Get user's token balance
    const tokenBalance = await tokenContract.balanceOf(user.walletAddress);
    const formattedBalance = formatUnits(tokenBalance, decimals);
    
    if (parseFloat(formattedBalance) <= 0) {
      console.log(`   User has no ${tradeInfo.tokenAddress} tokens to sell`);
      return null;
    }
    
    // Calculate percentage of holdings to sell (match the target's percentage)
    // This would require knowing the target's total holdings, which we don't have
    // For now, we'll just sell the same percentage of our balance as the target did
    // Assuming the target sold 100% for simplicity
    const sellAmount = formattedBalance;
    
    console.log(`   Mirroring sell: ${sellAmount} tokens of ${tradeInfo.tokenAddress}`);
    
    // If in sandbox mode, just simulate the trade
    if (copyTrade.sandboxMode) {
      console.log(`   Sandbox mode: Simulating trade`);
      
      // Get estimated ETH value
      const priceQuote = await getTokenPrice(tradeInfo.tokenAddress);
      const estimatedEth = parseFloat(sellAmount) * parseFloat(priceQuote.price);
      
      // Record simulated transaction
      await TransactionOps.createTransaction({
        telegramId: user.telegramId,
        type: 'simulated_sell',
        tokenAddress: tradeInfo.tokenAddress,
        amount: sellAmount,
        ethValue: estimatedEth.toString(),
        status: 'simulated',
        timestamp: new Date()
      });
      
      // Emit event for notification
      copyTradeEvents.emit('mirror_trade_simulated', {
        telegramId: user.telegramId,
        type: 'sell',
        tokenAddress: tradeInfo.tokenAddress,
        tokenAmount: sellAmount,
        estimatedEth: estimatedEth.toString()
      });
      
      return {
        simulated: true,
        tokenAmount: sellAmount,
        estimatedEth: estimatedEth.toString()
      };
    }
    
    // Execute real trade
    const privateKey = user.encryptedPrivateKey; // In a real implementation, this should be decrypted
    
    const result = await sellTokensForEth(
      tradeInfo.tokenAddress,
      sellAmount,
      privateKey,
      copyTrade.slippage,
      user.telegramId
    );
    
    // Emit event for notification
    copyTradeEvents.emit('mirror_trade_executed', {
      telegramId: user.telegramId,
      type: 'sell',
      tokenAddress: tradeInfo.tokenAddress,
      tokenAmount: sellAmount,
      ethAmount: result.ethAmount,
      txHash: result.txHash
    });
    
    return result;
  } catch (error) {
    console.error(`❌ Error executing mirror sell:`, error.message);
    return null;
  }
}

/**
 * Add a copy trade configuration
 * @param {string} telegramId - Telegram user ID
 * @param {string} targetWallet - Target wallet address
 * @param {string} maxEthPerTrade - Maximum ETH per trade
 * @param {number} slippage - Slippage percentage
 * @param {boolean} sandboxMode - Sandbox mode (simulate trades)
 * @returns {Promise<Object>} - Created copy trade configuration
 */
async function addCopyTrade(telegramId, targetWallet, maxEthPerTrade, slippage = 2.0, sandboxMode = true) {
  try {
    // Normalize wallet address
    const normalizedAddress = targetWallet.toLowerCase();
    
    // Create copy trade in database
    const copyTrade = await CopyTradeOps.createCopyTrade({
      telegramId,
      targetWallet: normalizedAddress,
      maxEthPerTrade,
      slippage,
      sandboxMode,
      active: true
    });
    
    // Start monitoring if not already active
    await startMonitoring(normalizedAddress);
    
    return copyTrade;
  } catch (error) {
    console.error(`❌ Error adding copy trade:`, error.message);
    throw error;
  }
}

/**
 * Update a copy trade configuration
 * @param {string} id - Copy trade ID
 * @param {Object} updateData - Data to update
 * @returns {Promise<boolean>} - True if updated successfully
 */
async function updateCopyTrade(id, updateData) {
  try {
    // Update copy trade in database
    const success = await CopyTradeOps.updateCopyTrade(id, updateData);
    return success;
  } catch (error) {
    console.error(`❌ Error updating copy trade:`, error.message);
    throw error;
  }
}

/**
 * Toggle copy trade active status
 * @param {string} id - Copy trade ID
 * @param {boolean} active - Active status
 * @returns {Promise<boolean>} - True if updated successfully
 */
async function toggleCopyTradeActive(id, active) {
  try {
    // Update copy trade in database
    const success = await CopyTradeOps.toggleCopyTradeActive(id, active);
    
    // If deactivating, check if we need to stop monitoring
    if (!active) {
      const copyTrade = await CopyTradeOps.getCopyTrade(id);
      
      if (copyTrade) {
        // Check if there are other active copy trades for this target
        const activeCopyTrades = await CopyTradeOps.getActiveTargetCopyTrades(copyTrade.targetWallet);
        
        if (activeCopyTrades.length === 0) {
          // No more active copy trades for this target, stop monitoring
          await stopMonitoring(copyTrade.targetWallet);
        }
      }
    }
    
    return success;
  } catch (error) {
    console.error(`❌ Error toggling copy trade active status:`, error.message);
    throw error;
  }
}

/**
 * Delete a copy trade configuration
 * @param {string} id - Copy trade ID
 * @returns {Promise<boolean>} - True if deleted successfully
 */
async function deleteCopyTrade(id) {
  try {
    // Get copy trade details first
    const copyTrade = await CopyTradeOps.getCopyTrade(id);
    
    if (!copyTrade) {
      return false;
    }
    
    // Delete copy trade from database
    const success = await CopyTradeOps.deleteCopyTrade(id);
    
    // Check if we need to stop monitoring
    if (success) {
      // Check if there are other active copy trades for this target
      const activeCopyTrades = await CopyTradeOps.getActiveTargetCopyTrades(copyTrade.targetWallet);
      
      if (activeCopyTrades.length === 0) {
        // No more active copy trades for this target, stop monitoring
        await stopMonitoring(copyTrade.targetWallet);
      }
    }
    
    return success;
  } catch (error) {
    console.error(`❌ Error deleting copy trade:`, error.message);
    throw error;
  }
}

/**
 * Get copy trade event emitter
 * @returns {EventEmitter} - Copy trade event emitter
 */
function getEventEmitter() {
  return copyTradeEvents;
}

module.exports = {
  initialize,
  startMonitoring,
  stopMonitoring,
  addCopyTrade,
  updateCopyTrade,
  toggleCopyTradeActive,
  deleteCopyTrade,
  getEventEmitter
}; 